File: 프로그래머스/lv2/81302. 거리두기 확인하기/거리두기 확인하기.py
from collections import deque

def solution(places):
    
    dx = (0, 0, -1, 1)
    dy = (-1, 1, 0, 0)
    
    def check(x, y):
        queue = deque([]); queue.append((x, y))
        visit = [[False] * 5 for _ in range(5)]; visit[x][y] = True
        
        while queue:
            cx, cy = queue.popleft()
            for i in range(4):
                nx = cx + dx[i]; ny = cy + dy[i]
                if 0 <= nx < 5 and 0 <= ny < 5:
                    if abs(x - nx) + abs(y - ny) <= 2: 
                        if not visit[nx][ny]:
                            if board[nx][ny] == 'P': return False
                            elif board[nx][ny] == 'X': continue
                            else: queue.append((nx, ny)); visit[nx][ny] = True
                    
        return True
                    
    
    result = []
    for place in places:
        
        board = []
        for p in place:
            board.append(list(p))
            
        break_flag = False
        for i in range(5):
            for j in range(5):
                if board[i][j] == 'P':
                    if not check(i, j): result.append(0); break_flag = True; break
            if break_flag: break
        
        if break_flag: continue
        else: result.append(1)
        
    return result

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.cpp
#include <string>
#include <vector>
#include <deque>
#include <tuple>

using namespace std;

int dx[4] = {0, 0, -1, 1};
int dy[4] = {-1, 1, 0, 0};

vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    
    for (auto place : places)
    {
        bool flag = false;
        for (int i = 0; i < 5; ++i)
        {
            for (int j = 0; j < 5; ++j)
            {
                if (place[i][j] == 'P')
                {
                    vector<vector<bool>> visit(5, vector<bool>(5, false)); visit[i][j] = true;
                    deque<tuple<int, int, int>> queue; queue.push_back({i, j, 0});
                    
                    while (!queue.empty())
                    {
                        auto [x, y, cnt] = queue.front(); queue.pop_front();
                        
                        if ((x != i || y != j) && (place[x][y] == 'P' && cnt <= 2))
                        {
                            flag = true;
                            break;
                        }
                        
                        for (int d = 0; d < 4; ++d)
                        {
                            int nx = x + dx[d]; int ny = y + dy[d];
                            if (0 <= nx && nx < 5 && 0 <= ny && ny < 5)
                            {
                                if (!visit[nx][ny] && place[nx][ny] != 'X')
                                {
                                    visit[nx][ny] = true;
                                    queue.push_back({nx, ny, cnt+1});
                                }
                            }
                        }
                    }
                }
                if (flag)
                {
                    break;
                }
            }
            if (flag)
            {
                break;
            }
        }
        if (flag)
        {
            answer.push_back(0);
        }
        else
        {
            answer.push_back(1);
        }
    }
    
    return answer;
}